### ISSUE TITLE
Integrate automatic differentiation with XLA conversion

### ISSUE THREAD

**Issue by Ebanflo42** [BradenEverson/unda#22]
My best case scenario for the next two weeks is that we get basic SGD up and running for simple dense networks (using XLA). The first major hurdle for this is not just differentiating the whole compute graph but identifying which parameter nodes the user wants gradients for. After that we add bindings to more XLA operations like matrix multiplication and activation functions + specify their differentiation rules, that is simple. Finally we should talk a bit about optimizers, but that is for another issue.

xla-rs defines "constants" and "parameters" in a pretty intuitive way: constants are baked into the executable whereas parameters are specified at every execution. Our current setup with the compute graphs reflects this design, which I'm happy with. The thing is, not everything which is specified at every execution should be differentiated. I saw Braden's email to ro, and this is highlighted by the question of how we distinguish "input", which is not differentiated but does change at every XLA call.

I would personally appreciate the following sort of API design (but haven't thought it through entirely):

```
let mut ctx = Context::new();
let learning_rate = ctx.scalar(-0.001);
let x = ctx.parameter("x", SmallVec::new()); // think of this as neural network weights
let y = ctx.parameter("y", SmallVec::new()); // think of this as neural network inputs
let loss = ctx.add(ctx.mul(x, x), y); // meaningless computation but that's not the point
let grad = ctx.gradient(x); // not yet implemented, this just gets the derivative with respect to x
let updates = ctx.mul(learning_rate, grad); // loss does not depend on this
// figure out how to apply updates when we talk about optimizers
let xla_executable = ctx.compile();
```

My idea is that `ctx.gradient` creates a node in the compute graph which is the placeholder for the gradient of a parameter. I think there's an issue with how the backend is currently set up relative to this: `compile` basically expects the loss node of the compute graph, and certainly this makes sense for `autodiff`, but overall we want the computation to potentially encapsulate computations on which the loss does not depend (note that updates depend on loss, not the other way around, so passing `ctx.compile(loss)` as we have it now would miss the multiplication of learning rate by the gradient). In summary, `compile` should take a loss node but not view it as the endpoint of the computation, instead it should use `autodiff` to connect placeholder `Gradient` nodes to actual computations which can then be converted to XLA.

This opens up the question of how the user specifies what tensors the XLA executable should return (usually you are interested in network predictions, loss, parameter updates, and maybe some auxiliary data), I will sleep on this. Interested in you guys' thoughts.
